tests/kernel/mem_protect/userspace/src/inherit.c

```diff
@@ -0,0 +1,105 @@
+/* 
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <ztest.h>
+#include <kernel_structs.h>
+#include <string.h>
+#include <stdlib.h>
+
+#define INHERIT_STACK_SIZE 1024
+
+/* function prototypes */
+static inline void dummy_start(struct k_timer *timer){ ARG_UNUSED(timer);}
+static inline void dummy_end(struct k_timer *timer){ ARG_UNUSED(timer);}
+
+/* Kernel objects */
+K_THREAD_STACK_DEFINE(test_1_stack, INHERIT_STACK_SIZE);
+K_SEM_DEFINE(inherit_sem, 0, 10);
+K_MUTEX_DEFINE(inhert_mutex);
+K_TIMER_DEFINE(inherti_timer, dummy_start, dummy_end);
+K_MSGQ_DEFINE(inherit_msgq, 10, 10, 2);
+__kernel struct k_thread test_1_tid;
+extern bool valid_fault;
+
+/* variables */
+#ifdef CONFIG_ARM
+u8_t __aligned(32) inherit_buf[32]; /* for mem domain */
+#else
+u8_t __aligned(4096) inherit_buf[4096]; /* for mem domain */
+#endif
+
+K_MEM_PARTITION_DEFINE(inherit_memory_partition, inherit_buf, sizeof(inherit_buf),
+                       K_MEM_PARTITION_P_RW_U_RW);
+
+struct k_mem_partition *inherit_memory_partition_array[] = {
+	&inherit_memory_partition
+};
+
+struct k_mem_domain inherit_mem_domain;
+
+__attribute__((optimize("O0"))) void access_test()
+{
+	u32_t msg_q_data = 0xA5A5;
+	/* check for all accesses  */
+	k_sem_give(&inherit_sem);
+	k_mutex_lock(&inhert_mutex,K_FOREVER);
+	(void) k_timer_status_get(&inherti_timer);
+	k_msgq_put(&inherit_msgq, (void *)&msg_q_data, K_NO_WAIT);
+	k_mutex_unlock(&inhert_mutex);
+	inherit_buf[10] = 0xA5;
+}
+
+void test_thread_1_for_user(void *p1, void *p2, void *p3)
+{
+	access_test();
+	if(_is_user_context()) {
+		/* 2nd time it comes here it should raise the test case as pass. */
+		ztest_test_pass();
+	}
+	/* k_thread_abort(k_current_get()); */
+	/* exit tc gracefully */
+}
+
+void test_thread_1_for_SU(void *p1, void *p2, void *p3)
+{
+	valid_fault = false;
+	access_test();
+	//k_mem_domain_remove_thread(k_current_get());
+	/* if no protection fault has occurred then supervisor mode inherit is working*/
```

Comment:
Do we really need to test the SU case in this testsuite?  This is supposed to be a testsuite of security properties enforced over userspace.

Reply:
That is just to confirm that that kobjects are clean and valid. I can also pass a null for the semaphore and still cause a fault. which is not the idea of this testing. This sets a base line, so that the next set of faults that come are genuine.